src/cmds/restic/cmd_find.go

```diff
@@ -84,7 +85,95 @@ func parseTime(str string) (time.Time, error) {
 	return time.Time{}, errors.Fatalf("unable to parse time: %q", str)
 }
 
-func findInTree(repo *repository.Repository, pat findPattern, id restic.ID, prefix string, snapshotID *string) error {
+type findNode restic.Node
+type omit *struct{}
+
+type statefulOutput struct {
+	ListLong bool
+	JSON     bool
+	inuse    bool
+	newsn    *restic.Snapshot
+	oldsn    *restic.Snapshot
+	hits     int
+}
+
+func (s *statefulOutput) PrintJSON(prefix string, node *restic.Node) {
+	if !s.inuse {
+		Printf("[")
+		s.inuse = true
+	}
+	if s.newsn != s.oldsn {
+		if s.oldsn != nil {
+			Printf("],\"hits\":%d,\"snapshot\":%q},", s.hits, s.oldsn.ID())
+		}
+		Printf(`{"matches":[`)
+		s.oldsn = s.newsn
+		s.hits = 0
+	}
+	if s.hits > 0 {
+		Printf(",")
+	}
+	b, err := json.Marshal(struct {
+		// Add these attributes
+		Path        string `json:"path,omitempty"`
+		Permissions string `json:"permissions,omitempty"`
+
+		*findNode
+
+		// Make the following attributes disappear
+		Name               omit `json:"name,omitempty"`
+		Inode              omit `json:"inode,omitempty"`
+		ExtendedAttributes omit `json:"extended_attributes,omitempty"`
+		Device             omit `json:"device,omitempty"`
+		Content            omit `json:"content,omitempty"`
+		Subtree            omit `json:"subtree,omitempty"`
+	}{
+		Path:        filepath.Join(prefix, node.Name),
+		Permissions: node.Mode.String(),
+		findNode:    (*findNode)(node),
+	})
+	if err != nil {
+		return
```

Comment:
Before this return it'd be nice to finalize the JSON string we already printed, so that even in the case of an error, the output is valid JSON (at least).

In addition you're silently ignoring an error here, use `Warnf()` to print it to stderr instead.